### PR TITLE
Add Event Log dumping

### PR DESCRIPTION
**Pull request by kuleshovilya**
Add Event Log dump to a log folder with a diagnostic files.
To test this PR, run any job and check the `_diag` folder in the agent layout. There should be an event log file there.

### CODE REVIEW

File: src/Agent.Worker/DiagnosticLogManager.cs [hunk 1]
```diff
@@ -346,5 +353,37 @@ private string GetEnvironmentContentNonWindows(int agentId, string agentName, IL
 
             return builder.ToString();
         }
+
+        private async Task<string> DumpEventLogs(DateTime jobStartTimeUtc, string diagFolder){
+            var builder = new StringBuilder();
+
+            string powerShellExe = HostContext.GetService<IPowerShellExeUtil>().GetPath();
```
Comment (anatolybolshakov) [inline src/Agent.Worker/DiagnosticLogManager.cs:360]: Let's catch exceptions here and print error message - so any issues with logs dumping won't break the agent itself.

File: src/Agent.Worker/DiagnosticLogManager.cs [hunk 2]
```diff
@@ -346,5 +353,37 @@ private string GetEnvironmentContentNonWindows(int agentId, string agentName, IL
...
+            string arguments = $@"Get-WinEvent -ListLog * | ForEach-Object {{ Get-WinEvent -ErrorAction SilentlyContinue -FilterHashtable @{{ LogName=$_.LogName; StartTime='{ jobStartTimeUtc }'; EndTime='{ DateTime.UtcNow }';}} }} | Format-List > { diagFolder }\EventViewer-{ jobStartTimeUtc.ToString("yyyyMMdd-HHmmss") }.log";
```
Comment (anatolybolshakov) [inline src/Agent.Worker/DiagnosticLogManager.cs:361]: One-line script here looks hard-to-read and modify - let's split it up a little bit.

Comment (DaniilShmelev) [reply src/Agent.Worker/DiagnosticLogManager.cs]: Done

File: src/Agent.Worker/DiagnosticLogManager.cs (hunk 1 above)
Comment (DaniilShmelev) [reply src/Agent.Worker/DiagnosticLogManager.cs]: Done

File: src/Agent.Worker/DiagnosticLogManager.cs [hunk 3]
```diff
@@ -105,6 +105,13 @@ public sealed class DiagnosticLogManager : AgentService, IDiagnosticLogManager
                 File.Copy(agentLogFile, destination);
             }
 
+            // Copy event logs for windows machines
+            if (PlatformUtil.RunningOnWindows)
+            {
+                executionContext.Debug("Dumping event viewer logs.");
```
Comment (anatolybolshakov) [inline src/Agent.Worker/DiagnosticLogManager.cs:111]: Nit: could it make sense to clarify that this is specifically for the current job - in the message and method name?

**Comment by anatolybolshakov** [comment]
Let's make some buddy testing also

**Comment by AndreyIvanov42** [comment]
I think it is needed to include generated log in the archive

File: src/Agent.Worker/DiagnosticLogManager.cs [hunk 4]
```diff
@@ -346,5 +353,33 @@ private string GetEnvironmentContentNonWindows(int agentId, string agentName, IL
...
+                        environment: null,
+                        requireExitCodeZero: false,
+                        outputEncoding: null,
+                        killProcessOnCancel: false,
+                        cancellationToken: default(CancellationToken));
+                }
+            }
+            catch
+            {
+                executionContext.Debug("Failed to dump event viewer logs. Skipping.");
```
Comment (AndreyIvanov42) [inline src/Agent.Worker/DiagnosticLogManager.cs:381]: I think that adding information about the error will be useful

Comment (DaniilShmelev) [reply src/Agent.Worker/DiagnosticLogManager.cs:386]: Done

**Comment by AndreyIvanov42** [comment]
LGTM!
During diagnostics, events logs are added to the archive. Tested on Windows 10.

**Review by AndreyIvanov42 — APPROVED** [review]

**Review by AndreyIvanov42 — APPROVED** [review]

File: src/Agent.Worker/DiagnosticLogManager.cs [hunk 5]
```diff
@@ -439,5 +451,34 @@ private string GetEnvironmentContentNonWindows(int agentId, string agentName, IL
 
             return builder.ToString();
         }
+
+        private async Task DumpCurrentJobEventLogs(IExecutionContext executionContext, string logFile, DateTime jobStartTimeUtc)
```
Comment (anatolybolshakov) [inline src/Agent.Worker/DiagnosticLogManager.cs:455]: Nit: please add comments to new methods

File: src/Agent.Worker/DiagnosticLogManager.cs [hunk 6]
```diff
@@ -148,6 +148,18 @@ public sealed class DiagnosticLogManager : AgentService, IDiagnosticLogManager
...
+            // Copy event logs for windows machines
+            if (PlatformUtil.RunningOnWindows)
+            {
+                executionContext.Debug("Dumping event viewer logs for current job.");
+
+                string eventLogsFile = $"{HostContext.GetDirectory(WellKnownDirectory.Diag)}/EventViewer-{ jobStartTimeUtc.ToString("yyyyMMdd-HHmmss") }.log";
+                await DumpCurrentJobEventLogs(executionContext, eventLogsFile, jobStartTimeUtc);
+
+                string destination = Path.Combine(supportFilesFolder, Path.GetFileName(eventLogsFile));
+                File.Copy(eventLogsFile, destination);
```
Comment (anatolybolshakov) [inline src/Agent.Worker/DiagnosticLogManager.cs:160]: What happens if file copying fails? Could it make sense to wrap all this logic with try..catch?

File: src/Agent.Worker/DiagnosticLogManager.cs (hunk 5 above)
Comment (DaniilShmelev) [reply src/Agent.Worker/DiagnosticLogManager.cs:465]: Done

File: src/Agent.Worker/DiagnosticLogManager.cs (hunk 6 above)
Comment (DaniilShmelev) [reply src/Agent.Worker/DiagnosticLogManager.cs]: If file copy fails, then the entire logs archive won't be uploaded to ADO. Wrapped it all in try-catch, thanks!

**Review by anatolybolshakov — APPROVED** [review]

**Pull request merged**